core/src/main/java/org/infinispan/persistence/support/ComposedSegmentedLoadWriteStore.java

```diff
@@ -198,16 +197,17 @@ public void deleteBatch(Iterable<Object> keys) {
 
    @Override
    public CompletionStage<Void> bulkUpdate(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
-      CompletableFuture<Void> future = new CompletableFuture<>();
-
-      Flowable.fromPublisher(publisher)
+      return Flowable.fromPublisher(publisher)
             .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
             .flatMap(groupedFlowable ->
                   groupedFlowable
                         .buffer(configuration.maxBatchSize())
-                        .doOnNext(batch -> stores.get(groupedFlowable.getKey()).bulkUpdate(Flowable.fromIterable(batch))))
-            .subscribe(Functions.emptyConsumer(), future::completeExceptionally, () -> future.complete(null));
-      return future;
+                        .flatMap(batch -> {
+                           CompletionStage<Void> stage = stores.get(groupedFlowable.getKey()).bulkUpdate(Flowable.fromIterable(batch));
+                           return RxJavaInterop.<Void>completionStageToPublisher().apply(stage);
```

Comment:
@wburns @ryanemerson there's no `@since` tag on `bulkUpdate`, but AFAICT it was only introduced in 10.0, so couldn't we change it to return a `Publisher` and avoid all this?

Reply:
To keep in line with all the other API requirements, the method should be returning `CompletionStage` as it is. The method has no actual return values and just needs a single notification that the updates are completed. Thus it should stay as is imo.

Reply:
I know what you mean, but it's annoying to see every store do the same `flowableToCompletionStage` conversion, sometimes using the helper and sometimes inlined, and especially here we have to go back and forth between publisher and completion stage. I guess we don't intend to keep using `ComposedSegmentedLoadWriteStore` forever though, at some point we'll convert SoftIndexFileStoreConfiguration and drop SingleFileStoreConfiguration :)